libcrux-psq/src/protocol/initiator.rs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    pub fn read_response(self, responder_msg: &Message) -> ResponderQueryPayload {
+        let Self {
+            responder_longterm_ecdh_pk,
+            initiator_ephemeral_ecdh_sk,
+            tx0,
+            k0,
+        } = self;
+        let tx2 = tx2(&tx0, &responder_msg.ephemeral_ecdh_pk);
+
+        let k2 = derive_k2_query_initiator(
+            &k0,
+            &responder_msg.ephemeral_ecdh_pk,
+            &initiator_ephemeral_ecdh_sk,
+            &responder_longterm_ecdh_pk,
+            &tx2,
+        );
+
+        k2.decrypt_deserialize(&responder_msg.ciphertext, &responder_msg.aad)
+    }
+}
+
+impl<'keys> RegistrationInitiatorPre<'keys> {
+    pub fn registration_message(
+        initiator_longterm_ecdh_pk: &'keys PublicKey,
+        initiator_longterm_ecdh_sk: &'keys PrivateKey,
+        responder_pq_pk: Option<&'keys MlKem768PublicKey>,
+        responder_longterm_ecdh_pk: &'keys PublicKey,
+        ctx: &[u8],
+        aad_outer: &[u8],
+        aad_inner: &[u8],
+        rng: &mut impl CryptoRng,
+    ) -> (Self, Message) {
+        let initiator_ephemeral_ecdh_sk = PrivateKey::new(rng);
+        let initiator_ephemeral_ecdh_pk = PublicKey::from(&initiator_ephemeral_ecdh_sk);
+
+        let (tx0, k0) = derive_k0(
+            responder_longterm_ecdh_pk,
+            &initiator_ephemeral_ecdh_pk,
+            &initiator_ephemeral_ecdh_sk,
+            ctx,
+            false,
+        );
+
+        let (pq_encaps, tx1, k1, ciphertext_inner) = encrypt_inner_payload(
+            initiator_longterm_ecdh_pk,
+            initiator_longterm_ecdh_sk,
+            responder_pq_pk,
+            responder_longterm_ecdh_pk,
+            aad_inner,
+            rng,
+            tx0,
+            &k0,
+        );
+
+        let payload0 = InitiatorOuterPayload::Registration {
+            pq_encaps,
+            aad: aad_inner.to_vec(),
+            ciphertext: ciphertext_inner,
+            initiator_longterm_ecdh_pk: initiator_longterm_ecdh_pk.clone(),
+        };
+
+        let ciphertext_outer = k0.serialize_encrypt(&payload0, &aad_outer);
+
+        (
+            Self {
+                responder_longterm_ecdh_pk,
+                initiator_ephemeral_ecdh_sk,
+                initiator_longterm_ecdh_sk,
+                initiator_longterm_ecdh_pk,
+                responder_pq_pk,
+                tx1,
+                k1,
+            },
+            Message {
+                ephemeral_ecdh_pk: initiator_ephemeral_ecdh_pk,
+                ciphertext: ciphertext_outer,
+                aad: aad_outer.to_vec(),
+            },
+        )
+    }
+
+    pub fn complete_registration(self, responder_msg: &'keys Message) -> SessionState<'keys> {
+        let Self {
+            responder_longterm_ecdh_pk,
+            initiator_ephemeral_ecdh_sk,
+            initiator_longterm_ecdh_sk,
+            initiator_longterm_ecdh_pk,
+            responder_pq_pk,
+            tx1,
+            k1,
+        } = self;
```

Comment:
This style is always pretty noisy. Is there a reason you don't just access the fields on `self`?